### DISCUSSION TITLE
Measure rendering time to debug complex layers that make the map too slow for the end user

### DISCUSSION THREAD

**Discussion by laem** [maplibre/maplibre-gl-js#5077] [Q&A]
## User Story

As a maplibre user, I'd want to measure the rendering times of each layer of my map style. 

## Rationale

 - Some map styles render quicker than others. Generating the good tiles compromise is not easy. Measure tile request data is one thing, but measuring rendering times is closer to what burdens my users' experience 
 - I'm developping cartes.app, an alternative to Google Maps, and I want my maps to be fast enough for devices with not-so powerful GPUs
 - I haven't found any way to do it today

**Comment by HarelM** [comment]
There are some performance indicators sprinkled in the code, you might want check those.
There's also this PR, which is talking about performance calculation and defintions
- https://github.com/maplibre/maplibre-gl-js/pull/4901

And there's also the benchmark code in this repo you might be able to leverage to measure stuff.
If all those tools don't help (which I tend to think will be the case unfortunately), you can suggest a way to better measure things.

As a side note, if you need to support weak GPUs MapLibre might not be the right solution as it relies heavily on the GPU, even for raster maps, so it might be beneficial to use a different library that doesn't utilize the GPU as much such as leaflet or openlayers.

**Reply by laem** [reply]
Thanks, I'll look into these solutions.
